File: push-be/internal/infra/oauth.go
package infra

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"push-be/internal/domain/usecase"
)

type HTTPOAuthClient struct {
	http *http.Client
}

func NewHTTPOAuthClient() *HTTPOAuthClient {
	return &HTTPOAuthClient{http: &http.Client{Timeout: 15 * time.Second}}
}

func (c *HTTPOAuthClient) ExchangeCode(ctx context.Context, cfg usecase.ProviderConfig, code, redirectURI string) (string, error) {
	form := url.Values{}
	form.Set("client_id", cfg.ClientID)
	form.Set("client_secret", cfg.ClientSecret)
	form.Set("code", code)
	form.Set("redirect_uri", redirectURI)
	form.Set("grant_type", "authorization_code")
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, cfg.TokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	resp, err := c.http.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	var body struct {
		AccessToken string `json:"access_token"`
		Error       string `json:"error"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return "", err
	}
	if resp.StatusCode != http.StatusOK || body.AccessToken == "" {
		return "", fmt.Errorf("token exchange failed: status %d %s", resp.StatusCode, body.Error)
	}
	return body.AccessToken, nil
}

func (c *HTTPOAuthClient) FetchIdentity(ctx context.Context, cfg usecase.ProviderConfig, accessToken string) (string, string, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cfg.UserInfoURL, nil)
	if err != nil {
		return "", "", err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Accept", "application/json")
	resp, err := c.http.Do(req)
	if err != nil {
		return "", "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return "", "", errors.New("identity fetch failed")
	}
	var body struct {
		Sub   string `json:"sub"`
		ID    int64  `json:"id"`
		Name  string `json:"name"`
		Login string `json:"login"`
		Email string `json:"email"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return "", "", err
	}
	subject := body.Sub
	if subject == "" && body.ID != 0 {
		subject = strconv.FormatInt(body.ID, 10)
	}
	if subject == "" {
		return "", "", errors.New("missing subject")
	}
	name := body.Name
	if name == "" {
		name = body.Login
	}
	if name == "" {
		name = body.Email
	}
	if name == "" {
		name = "user"
	}
	return subject, name, nil
}

File: push-be/app/infrastructure/oauth.py
from __future__ import annotations
from dataclasses import dataclass

import httpx


@dataclass
class ProviderConfig:
    client_id: str
    client_secret: str
    auth_url: str
    token_url: str
    user_info_url: str


class OAuthClient:
    def __init__(self):
        self._client = httpx.AsyncClient(timeout=15)

    async def exchange_code(self, cfg: ProviderConfig, code: str, redirect_uri: str) -> str:
        resp = await self._client.post(cfg.token_url, data={
            "client_id": cfg.client_id,
            "client_secret": cfg.client_secret,
            "code": code,
            "redirect_uri": redirect_uri,
            "grant_type": "authorization_code",
        }, headers={"Accept": "application/json"})
        body = resp.json() if resp.content else {}
        if resp.status_code != 200 or not body.get("access_token"):
            raise ValueError(f"token exchange failed: status {resp.status_code} {body.get('error', '')}")
        return body["access_token"]

    async def fetch_identity(self, cfg: ProviderConfig, access_token: str) -> tuple[str, str]:
        resp = await self._client.get(cfg.user_info_url, headers={
            "Authorization": f"Bearer {access_token}",
            "Accept": "application/json",
        })
        if resp.status_code != 200:
            raise ValueError("identity fetch failed")
        body = resp.json()
        subject = body.get("sub") or (str(body["id"]) if body.get("id") else "")
        if not subject:
            raise ValueError("missing subject")
        name = body.get("name") or body.get("login") or body.get("email") or "user"
        return subject, name

    async def aclose(self) -> None:
        await self._client.aclose()


def provider_configs(cfg) -> dict[str, ProviderConfig]:
    return {
        "google": ProviderConfig(
            client_id=cfg.google.client_id, client_secret=cfg.google.client_secret,
            auth_url="https://accounts.google.com/o/oauth2/v2/auth",
            token_url="https://oauth2.googleapis.com/token",
            user_info_url="https://openidconnect.googleapis.com/v1/userinfo"),
        "github": ProviderConfig(
            client_id=cfg.github.client_id, client_secret=cfg.github.client_secret,
            auth_url="https://github.com/login/oauth/authorize",
            token_url="https://github.com/login/oauth/access_token",
            user_info_url="https://api.github.com/user"),
    }
